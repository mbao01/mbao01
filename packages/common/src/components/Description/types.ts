export type DescriptionProps = React.HTMLAttributes<HTMLDListElement>;

export type DescriptionTermProps = React.HTMLAttributes<HTMLElement>;

export type DescriptionDetailProps = React.HTMLAttributes<HTMLElement>;
