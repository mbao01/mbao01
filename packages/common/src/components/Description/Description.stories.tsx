import type { Meta, StoryObj } from "@storybook/react-vite";
import { Text } from "../Text";
import { Description } from "./Description";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Description,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Description>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    children: (
      <>
        <Description.Term>Name</Description.Term>
        <Description.Detail>Jonny Bravo</Description.Detail>
      </>
    ),
  },
};

export const MultipleItemsInList: Story = {
  args: {
    children: (
      <>
        <Description.Term>Name</Description.Term>
        <Description.Detail>Jonny Bravo</Description.Detail>
        <Description.Term>Age</Description.Term>
        <Description.Detail>28</Description.Detail>
        <Description.Term>Profession</Description.Term>
        <Description.Detail>Sporting Coach</Description.Detail>
        <Description.Detail>
          <Text size="base" className="italic">
            Associate
          </Text>
        </Description.Detail>
      </>
    ),
  },
};
