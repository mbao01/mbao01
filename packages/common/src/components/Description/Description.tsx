import type { DescriptionDetailProps, DescriptionProps, DescriptionTermProps } from "./types";
import { cn } from "../../utilities";

const Description = (props: DescriptionProps) => {
  return <dl {...props} />;
};

const DescriptionTerm = ({ className, ...props }: DescriptionTermProps) => {
  return <dt className={cn("py-2 text-sm font-semibold", className)} {...props} />;
};

const DescriptionDetail = ({ className, ...props }: DescriptionDetailProps) => {
  return <dd className={cn("text-base not-last:mb-1", className)} {...props} />;
};

Description.Term = DescriptionTerm;
Description.Detail = DescriptionDetail;

export { Description };
