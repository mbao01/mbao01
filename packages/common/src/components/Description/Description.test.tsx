import { render, screen, within } from "@testing-library/react";
import { Description } from "./Description";

describe("Description", () => {
  it("description list with a single term", () => {
    const { asFragment } = render(
      <Description>
        <Description.Term>Name</Description.Term>
        <Description.Detail>Evelyn Gilts</Description.Detail>
      </Description>
    );

    expect(screen.getByRole("term")).toHaveTextContent("Name");
    expect(screen.getByRole("definition")).toHaveTextContent("Evelyn Gilts");
    expect(asFragment()).toMatchSnapshot();
  });

  it("description list with multiple terms", () => {
    const { asFragment } = render(
      <Description>
        <Description.Term>Name</Description.Term>
        <Description.Detail>Evelyn Gilts</Description.Detail>
        <Description.Term>Age</Description.Term>
        <Description.Detail>23</Description.Detail>
        <Description.Term>Profession</Description.Term>
        <Description.Detail>AI Specialist</Description.Detail>
      </Description>
    );
    const terms = ["Name", "Age", "Profession"];
    const details = ["Evelyn Gilts", "23", "AI Specialist"];

    const termsEls = screen.getAllByRole("term");
    const definitionEls = screen.getAllByRole("definition");
    termsEls.forEach((termEl, index) => {
      expect(termEl).toHaveTextContent(terms[index]);
      expect(definitionEls[index]).toHaveTextContent(details[index]);
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
