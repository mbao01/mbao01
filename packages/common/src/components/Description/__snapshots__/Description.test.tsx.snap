// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Description > description list with a single term 1`] = `
<DocumentFragment>
  <dl>
    <dt
      class="py-2 text-sm font-semibold"
    >
      Name
    </dt>
    <dd
      class="text-base not-last:mb-1"
    >
      Evelyn Gilts
    </dd>
  </dl>
</DocumentFragment>
`;

exports[`Description > description list with multiple terms 1`] = `
<DocumentFragment>
  <dl>
    <dt
      class="py-2 text-sm font-semibold"
    >
      Name
    </dt>
    <dd
      class="text-base not-last:mb-1"
    >
      Evelyn Gilts
    </dd>
    <dt
      class="py-2 text-sm font-semibold"
    >
      Age
    </dt>
    <dd
      class="text-base not-last:mb-1"
    >
      23
    </dd>
    <dt
      class="py-2 text-sm font-semibold"
    >
      Profession
    </dt>
    <dd
      class="text-base not-last:mb-1"
    >
      AI Specialist
    </dd>
  </dl>
</DocumentFragment>
`;
